src/Microsoft.VisualStudio.Threading/AsyncReaderWriterLock.cs

```diff
@@ -1718,10 +1818,35 @@ private void PendAwaiter(Awaiter awaiter)
                 {
                     Queue<Awaiter>? queue = this.GetLockQueue(awaiter.Kind);
                     queue.Enqueue(awaiter);
+
+                    if (awaiter.Kind == LockKind.Write)
+                    {
+                        this.EnablePendingWriterLockDeadLockWatching();
+                    }
                 }
             }
         }
 
+        private void EnablePendingWriterLockDeadLockWatching()
+        {
+            if (this.joinableTaskContext != null &&
+                this.pendingWriterLockDeadLockCheckTimer == null &&
+                this.waitingWriters.Count > 0 &&
+                (this.issuedReadLocks.Count > 0 || this.issuedUpgradeableReadLocks.Count > 0))
+            {
+                this.pendingWriterLockDeadLockCheckTimer = new Timer(PendingWriterLockDeadLockWatchingCallback, this, this.DeadLockCheckTimeOutInMilliseconds, -1);
+            }
+        }
+
+        private void StopPendingWriterLockDeadLockWatching()
+        {
+            if (this.pendingWriterLockDeadLockCheckTimer != null && this.waitingWriters.Count == 0)
```

Comment:
Why do we check `this.waitingWriters.Count == 0` and only do what the method name claims it will do if that condition is true? Shouldn't we *always* Stop if the method is called to stop it?
Should the caller check this condition before calling this method? Should we rename this method to "Stop...If..."?

Reply:
We actually checked it before calling the function. Yes, it makes sense to remove that part of check in this function.